src/compiler/checker.ts

```diff
@@ -28417,6 +28417,47 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     }
 
     function checkObjectLiteral(node: ObjectLiteralExpression, checkMode?: CheckMode): Type {
+        const contextualType = getContextualType(node, /*contextFlags*/ undefined);
+        const isContextualTypeDependent = 
+            contextualType &&
+            contextualType.immediateBaseConstraint && 
+            contextualType.immediateBaseConstraint.aliasTypeArguments &&
+            contextualType.immediateBaseConstraint.aliasTypeArguments.length === 1 &&
+            contextualType.immediateBaseConstraint.aliasTypeArguments[0].id === contextualType.id;
+
+        if (!isContextualTypeDependent) {
+            return checkObjectLiteralNonDependently(node, checkMode);
+        }
+
+        let valueType = checkObjectLiteralNonDependently(node, checkMode);
+
+        let newContextualType = cloneTypeParameter(contextualType);
+        newContextualType.immediateBaseConstraint =
+            instantiateType(
+                contextualType.immediateBaseConstraint,
+                createTypeMapper([contextualType], [valueType])
+            );
+        if (newContextualType.immediateBaseConstraint!.flags & TypeFlags.StructuredType) {
+            newContextualType.immediateBaseConstraint = resolveStructuredTypeMembers(newContextualType.immediateBaseConstraint as StructuredType);
+        }

```

Comment:
I don't know why this is needed, but given it is how do you resolve any kind of type and not just structured?